src/utils.h

```diff
@@ -377,5 +377,167 @@ Rescale(Eigen::ArrayXXd               weights,
   intercept_archive.emplace_back(std::move(intercept));
 }
 
+
+//' Return a vector of random index
+//' 
+//' @param n_samples the sampler uses range from 0 to n_samples
+//' @param length the number of random index
+
+Eigen::ArrayXXi
+IndexStochastic(const unsigned n_samples,
+                const unsigned length)
+{
+  Eigen::ArrayXXi index(1, length);
+  for (unsigned i = 0; i < length; ++i) {
+    unsigned s_ind = floor(R::runif(0.0, n_samples));
+    index.col(i) = s_ind;
+  }
+  return index;
+}
+
+//' Return a vector of cyclic index
+//' 
+//' @param n_samples the sampler uses range from 0 to n_samples
+//' @param length the number of random index
+
+Eigen::ArrayXXi
+IndexCyclic(const unsigned n_samples,
+            const unsigned length)
+{
+  Eigen::ArrayXXi index(1, length);
+  index.row(0) = Eigen::ArrayXi::LinSpaced(n_samples, 0, n_samples);
+  return index;
+}
+
+//' wrapper aroud R's RNG such that we get a unifrom distribution over 
+//' [0,n) as required by the STL algorithm
+inline int randWrapper(const int n) { return floor(unif_rand()*n); }
```

Comment:
Is there an `Rcpp` version we could use here instead? It's pretty hard to tell without going hunting that `unif_rand` is from R's C API.

Reply:
Hi Michael, it seems that c++ does not respect random number generators from R, so I am using this to let the results be controlled by setting the seed on the R side following Johan’s suggestion.

Reply:
I didn't mean use a C++ RNG: I still meant use the R RNGs via C++. 

Maybe  ```Rcpp::as<int>(Rcpp::sample(n, 1)) - 1```

My point was that it's not obvious (to me) that `unif_rand` does use the R RNGs, as opposed to being a C++ standard function or something from Eigen.